### PR TITLE
Vb 2722 get one off report showing session capacity and take up

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Session capacity reporting
## What is the intent behind these changes?

Adds feature to report on session capacity and take up on a daily basis

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    val sessionDetails = mutableMapOf<SessionScheduleDto, Map<Pair<VisitStatus, VisitRestriction>, Int>>()
+    sessions.forEach { session ->
+      sessionDetails[session] = getVisitCountsForSession(session, reportDate)
+    }
+
+    return sessionDetails
+  }
+
+  private fun getAllActivePrisons(): List<PrisonDto> {
+    return prisonConfigService.getPrisons().filter { it.active }
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt:65]: Should use prisonConfigService.getSupportedPrisons()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  private val prisonConfigService: PrisonConfigService,
+  private val sessionService: SessionService,
+  private val visitRepository: VisitRepository,
+) {
+  fun getVisitCountsBySession(reportDate: LocalDate): List<SessionVisitCountsDto> {
+    return getSessionsReport(reportDate)
+  }
+
+  private fun getSessionsReport(reportDate: LocalDate): List<SessionVisitCountsDto> {
+    val prisons = getAllActivePrisons()
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt:26]: Am I correct in thinking you are only interested in prison that are not exclude by the report date, if so I would create a new method in the prisonConfigService that returns prison that are valid for the report date.

**Review by AledWEvans — COMMENTED** [review]
Some comments have a look

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    visitCounts[Pair(VisitStatus.BOOKED, VisitRestriction.CLOSED)] = bookedCounts.firstOrNull { it.visitRestriction == VisitRestriction.CLOSED }?.count ?: 0
+    visitCounts[Pair(VisitStatus.CANCELLED, VisitRestriction.OPEN)] = cancelledCounts.firstOrNull { it.visitRestriction == VisitRestriction.OPEN }?.count ?: 0
+    visitCounts[Pair(VisitStatus.CANCELLED, VisitRestriction.CLOSED)] = cancelledCounts.firstOrNull { it.visitRestriction == VisitRestriction.CLOSED }?.count ?: 0
+
+    return visitCounts
+  }
+
+  private fun getVisitCountsBySession(sessionTemplateReference: String, visitStatus: VisitStatus, visitDate: LocalDate): List<VisitRestrictionStats> {
+    return when (visitStatus) {
+      VisitStatus.BOOKED -> {
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt:91]: For cancelled the SQL adds outcome_status != 'SUPERSEDED_CANCELLATION') so had to create a different method for cancel as getCountOfCancelledSessionVisitsForOpenOrClosedRestriction

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt:26]: I need all prisons but if the date is excluded or there are no sessions we are setting flag so we can filter by prison and use them on reports

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt:65]: getSupportedPrisons() returns a list of Strings while I need the PrisonDtos back. Should I add a new method on the PrisonConfigService as getActivePrisons(): List<PrisonDto>?

Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt:65]: I think that getSupportedPrisons prisons is :
  @Query(
    "select p.code from Prison p " +
      "where p.active = true order by p.code",
  )
  fun getSupportedPrisons(): List<String>

could call it differently

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt:91]: I was being blind,... different call different code :-)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/reporting/SessionVisitCountsDto.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import io.swagger.v3.oas.annotations.media.Schema
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.SessionCapacityDto
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.SessionTimeSlotDto
+import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitType
+import java.time.LocalDate
+
+data class SessionVisitCountsDto(
+  @Schema(description = "Date of Report", example = "2023-09-01", required = true)
+  val reportDate: LocalDate,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/reporting/SessionVisitCountsDto.kt:11]: no annotation to enforce required fields just swagger,... same for below attributes

Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/reporting/SessionVisitCountsDto.kt:12]: Thanks! Updated Dto and code based on the comments.

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**